server/blob/blob-union/src/main/java/org/apache/james/blob/union/HybridBlobStore.java

```diff
@@ -128,45 +145,53 @@ private HybridBlobStore(BlobStore lowCostBlobStore, BlobStore highPerformanceBlo
 
     @Override
     public Mono<BlobId> save(BucketName bucketName, byte[] data, StoragePolicy storagePolicy) {
-        return selectBlobStore(storagePolicy, Mono.just(data.length > configuration.getSizeThreshold()))
-            .flatMap(blobStore -> blobStore.save(bucketName, data, storagePolicy));
+        return save(bucketName, new ByteArrayInputStream(data), storagePolicy);
     }
 
     @Override
     public Mono<BlobId> save(BucketName bucketName, InputStream data, StoragePolicy storagePolicy) {
         Preconditions.checkNotNull(data);
 
-        BufferedInputStream bufferedInputStream = new BufferedInputStream(data, configuration.getSizeThreshold() + 1);
-        return selectBlobStore(storagePolicy, Mono.fromCallable(() -> isItABigStream(bufferedInputStream)))
-            .flatMap(blobStore -> blobStore.save(bucketName, bufferedInputStream, storagePolicy));
-    }
-
-    private Mono<BlobStore> selectBlobStore(StoragePolicy storagePolicy, Mono<Boolean> largeData) {
         switch (storagePolicy) {
             case LOW_COST:
-                return Mono.just(lowCostBlobStore);
+                return lowCostBlobStore.save(bucketName, data, storagePolicy);
+
             case SIZE_BASED:
-                return largeData.map(isLarge -> {
-                    if (isLarge) {
-                        return lowCostBlobStore;
-                    }
-                    return highPerformanceBlobStore;
-                });
-            case HIGH_PERFORMANCE:
-                return Mono.just(highPerformanceBlobStore);
+                BufferedInputStream bufferedInputStream = new BufferedInputStream(data, configuration.getSizeThreshold() + 1);
+                return Mono.fromCallable(() -> isItABigStream(bufferedInputStream))
+                    .filter(Boolean::booleanValue)
```

Comment:
Consider using Mono.using() for resource cleanup.

Reply:
Which resource we gonna clean btw?

Reply:
Java GC is safe enough to clean up the buffered input stream.

Closing the `data` inputStream needs to be done.

@nvduc91 in a separate commit can you add a test to ensure `blobStore.save` always closes the supplied inputStream (In BlobStoreContract, with a large and little stream)

Reply:
I tried and turnout it gonna a big refactoring.I proposed make it own PR.